cirq/contrib/quimb/mps_simulator.py

```diff
@@ -263,60 +336,98 @@ def __init__(self, qubit_map, rsum2_cutoff, initial_state=0):
-        end_bras = []
-        if skip_tracing_out_for_qubits:
-            for qubit, i in self.qubit_map.items():
-                if i not in skip_tracing_out_for_qubits:
-                    d = qubit.dimension
-                    data = [math.sqrt(d)] * d
-                    end_bras.append(qtn.Tensor(data=data, inds=(self.i_str(i),)))
-        tensor_network = qtn.TensorNetwork(self.M + end_bras)
-        return tensor_network.contract(inplace=False)
-
-    def partial_state_vector(self, skip_tracing_out_for_qubits):
-        M = self._sum_up(
-            skip_tracing_out_for_qubits={
-                self.qubit_map[qubit] for qubit in skip_tracing_out_for_qubits
-            }
-        )
+    def state_vector(self) -> np.ndarray:
+        """Returns the full state vector.
+
+        Returns:
+            A vector that contains the full state.
+        """
+        tensor_network = qtn.TensorNetwork(self.M)
+        state_vector = tensor_network.contract(inplace=False)
+
         # Here, we rely on the formatting of the indices, and the fact that we have enough
         # leading zeros so that 003 comes before 100.
-        sorted_ind = tuple(sorted(M.inds))
-        return M.fuse({'i': sorted_ind}).data
+        sorted_ind = tuple(sorted(state_vector.inds))
+        return state_vector.fuse({'i': sorted_ind}).data
+
+    def partial_trace(self, keep_qubits: Set[ops.Qid]) -> np.ndarray:
+        """Returns the partial trace only for the keep_qubits specified.
+
+        Args:
+            keep_qubits: The set of qubits that are left after computing the
+                partial trace. For example, if we have a circuit for 3 qubits
+                and this parameter only has one qubit, the entire density matrix
+                would be 8x8, but this function returns a 2x2 matrix.
+
+        Returns:
+            An array that contains the partial trace.
+        """
+
+        contracted_inds = set(
+            [self.i_str(i) for qubit, i in self.qubit_map.items() if qubit not in keep_qubits]
+        )
+
+        CONJ_PFX = "conj_"
 
-    def state_vector(self):
-        return self.partial_state_vector(skip_tracing_out_for_qubits={})
+        tensor_network = qtn.TensorNetwork(self.M)
+
+        # Rename the internal indices to avoid collisions. Also rename the qubit
+        # indices that are kept. We do not rename the qubit indices that are
+        # traced out.
+        conj_tensor_network = tensor_network.conj()
+        for M in conj_tensor_network.tensors:
+            for ind in M.inds:
+                if ind not in contracted_inds:
+                    conj_tensor_network.reindex({ind: CONJ_PFX + ind}, inplace=True)
+        partial_trace = conj_tensor_network @ tensor_network
+
+        old_inds = [self.i_str(self.qubit_map[keep_qubit]) for keep_qubit in keep_qubits]
+        new_inds = [CONJ_PFX + old_ind for old_ind in old_inds]
+        return partial_trace.to_dense(old_inds, new_inds)
```

Comment:
"old" and "new" are bad names for this. How about `orig_inds, conj_inds`. or `forward_inds, backward_inds`. I sortof leaned into the "forwards backwards" thing in the quimb density simulator as that's how I like to think about it: you keep two copies of the circuit and one runs "forwards" i.e. normal and the other applies the conj transpose of each operation, i.e. it evolves "backwards" in time

Reply:
Done.